attribute-defined-outside-init when assigning inside __post_init__

<!--
  Hi there! Thank you for discovering and submitting an issue.

  Before you submit this, make sure that the issue doesn't already exist
  or if it is not closed.

  Is your issue fixed on the preview release?: pip install pylint astroid --pre -U

-->

### Steps to reproduce
1. Create file `a.py`:
```
from dataclasses import dataclass, InitVar


@dataclass
class Class():
    var: InitVar[int]

    def __post_init__(self, var: int):
        self.other_var = var + 2
```

### Current behavior
`pylint a.py` produces
```
************* Module a
a.py:1:0: C0111: Missing module docstring (missing-docstring)
a.py:4:0: C0111: Missing class docstring (missing-docstring)
a.py:9:8: W0201: Attribute 'other_var' defined outside __init__ (attribute-defined-outside-init)

------------------------------------------------------------------
Your code has been rated at 4.00/10 (previous run: 4.00/10, +0.00)
```

### Expected behavior

All lines except the one with `W0201` should be printed because `@dataclass` creates a `__init__` method that calls `__post_init__`
### pylint --version output
```
pylint 2.1.1
astroid 2.0.4
Python 3.7.0 (default, Jun 28 2018, 08:04:48) [MSC v.1912 64 bit (AMD64)]
```

--- a/CONTRIBUTORS.txt
+++ b/CONTRIBUTORS.txt
@@ -305,3 +305,5 @@ contributors:
 * Agustin Toledo: contributor
 
 * Nicholas Smith: contributor
+
+* Andrzej Klajnert: contributor

--- a/ChangeLog
+++ b/ChangeLog
@@ -158,6 +158,10 @@ Release date: TBA
 
   Close #2963
 
+* Added ``__post_init__`` to ``defining-attr-methods`` in order to avoid ``attribute-defined-outside-init`` in dataclasses.
+
+  Close #2581
+
 What's New in Pylint 2.3.0?
 ===========================
 

--- a/examples/pylintrc
+++ b/examples/pylintrc
@@ -460,7 +460,8 @@ redefining-builtins-modules=six.moves,past.builtins,future.builtins,builtins,io
 # List of method names used to declare (i.e. assign) instance attributes.
 defining-attr-methods=__init__,
                       __new__,
-                      setUp
+                      setUp,
+                      __post_init__
 
 # List of member names, which should be excluded from the protected access
 # warning.

--- a/pylint/checkers/classes.py
+++ b/pylint/checkers/classes.py
@@ -666,7 +666,7 @@ class ClassChecker(BaseChecker):
         (
             "defining-attr-methods",
             {
-                "default": ("__init__", "__new__", "setUp"),
+                "default": ("__init__", "__new__", "setUp", "__post_init__"),
                 "type": "csv",
                 "metavar": "<method names>",
                 "help": "List of method names used to declare (i.e. assign) \

--- a/pylintrc
+++ b/pylintrc
@@ -334,7 +334,7 @@ max-public-methods=25
 [CLASSES]
 
 # List of method names used to declare (i.e. assign) instance attributes.
-defining-attr-methods=__init__,__new__,setUp
+defining-attr-methods=__init__,__new__,setUp,__post_init__
 
 # List of valid names for the first argument in a class method.
 valid-classmethod-first-arg=cls

--- a/tests/functional/attribute_defined_outside_init.py
+++ b/tests/functional/attribute_defined_outside_init.py
@@ -78,3 +78,7 @@ class Mine:
     @prop.setter
     def prop(self, value):
         self.__prop = value
+
+class DataClass:
+    def __post_init__(self):
+        self.a = 42

